src/Microsoft.DotNet.Interactive/RoutingSlip.cs

```diff
@@ -11,49 +12,103 @@ namespace Microsoft.DotNet.Interactive;
     {
         if (source is { })
         {
             _uniqueUris = new HashSet<Uri>(source);
-            _uris = new List<Uri>(source);
+            _uris = new List<Entry>(source._uris);
+            _entriesByUri = new ConcurrentDictionary<Uri, Entry>(source._entriesByUri);
         }
         else
         {
             _uniqueUris = new HashSet<Uri>();
-            _uris = new List<Uri>();
+            _entriesByUri = new ConcurrentDictionary<Uri, Entry>();
+            _uris = new List<Entry>();
         }
     }
 
-    public bool TryAdd(Uri kernelOrKernelHostUri)
+    public bool TryMarkArrival(Uri kernelOrKernelHostUri)
     {
-        lock (_lock)
+        return TryMark(kernelOrKernelHostUri, false);
+    }
+
+    public bool TryMarkHandled(Uri kernelOrKernelHostUri)
+    {
+        return TryMark(kernelOrKernelHostUri, true);
+    }
+
+    private bool TryMark(Uri kernelOrKernelHostUri, bool handled)
+    {
+        var entry = new Entry(kernelOrKernelHostUri)
         {
-            if (_uniqueUris.Add(kernelOrKernelHostUri))
+            Handled = handled
+        };
+        var ret = _entriesByUri.TryAdd(kernelOrKernelHostUri, entry);
+        if (ret)
+        {
+            _uniqueUris.Add(kernelOrKernelHostUri);
+            _uris.Add(new Entry(kernelOrKernelHostUri));
+        }
+
+        return ret;
+    }
+
+    public void MarkHandled(Uri kernelOrKernelHostUri)
+    {
+        if (_entriesByUri.TryGetValue(kernelOrKernelHostUri, out var entry))
+        {
+            if (entry.Handled)
+            {
+                throw new InvalidOperationException($"The routing slip has already been handled by {kernelOrKernelHostUri}");
+            }
+            else
             {
-                _uris.Add(kernelOrKernelHostUri);
-                return true;
+                entry.Handled = true;
             }
         }
-
-        return false;
+        else
+        {
+            throw new InvalidOperationException(
+                $"The kernel or kernel host {kernelOrKernelHostUri} is not part of this routing slip");
+        }
     }
 
-    public bool Contains(Uri kernelOrKernelHostUri) => _uniqueUris.Contains(kernelOrKernelHostUri);
+    public bool Contains(Uri kernelOrKernelHostUri) => _entriesByUri.ContainsKey(kernelOrKernelHostUri);
 
     public bool Contains(RoutingSlip other)
     {
-        var contains =  this.Zip(other, (o, i) => o.Equals(i)).All(x => x);
+        var contains_old = this.Zip(other, (o, i) => o.Equals(i)).All(x => x);
+        var contains = _uniqueUris.IsSupersetOf(other._uniqueUris);
         return contains;
     }
 
-    IEnumerator<Uri> IEnumerable<Uri>.GetEnumerator() => _uris.GetEnumerator();
+    IEnumerator<Uri> IEnumerable<Uri>.GetEnumerator() => _uris.Select(e => e.KernelUri).GetEnumerator();
 
     IEnumerator IEnumerable.GetEnumerator() => _uris.GetEnumerator();
 
-    public Uri this[int index] => _uris[index];
+    public Uri this[int index] => _uris[index].KernelUri;
```

Comment:
I don't think the indexer makes much sense anymore, and it probably no longer makes sense for the routing slip to be `IEnumerable<Uri>`.